src/java/org/apache/cassandra/index/sai/disk/SSTableComponentsWriter.java

```diff
@@ -76,65 +52,36 @@ default void abort(Throwable accumulator)
         }
 
         public void nextRow(PrimaryKey key) throws MemtableTrie.SpaceExhaustedException, IOException
         {
-            if (key.size() <= MAX_RECURSIVE_KEY_LENGTH)
-                rowMapping.putRecursive(v -> key.asComparableBytes(v), key.sstableRowId(), (existing, neww) -> neww);
-            else
-                rowMapping.apply(Trie.singleton(v -> key.asComparableBytes(v), key.sstableRowId()), (existing, neww) -> neww);
-            maxSSTableRowId = key.sstableRowId();
-            maxKeyLength = Math.max(maxKeyLength, key.asBytes().length);
-            rowCount++;
-            // If the trie is full then we need to flush it and start a new one
-            if (rowMapping.reachedAllocatedSizeThreshold())
-                flush();
+            long offset = primaryKeys.getFilePointer();
+            byte[] rawKey = key.asBytes();
+            primaryKeys.writeBytes(rawKey, rawKey.length);
+            primaryKeyOffsets.add(offset);
         }
 
         public void complete() throws IOException
         {
-            if (rowCount > 0)
-                flush();
-            maxKeyLength = partitionKeysMetas.stream().map(meta -> meta.bytesPerDim).max(Comparator.naturalOrder()).orElse(0);
-            rowCount = partitionKeysMetas.stream().mapToLong(meta -> meta.pointCount).sum();
-            List<BKDReader> readers = new ArrayList<>(partitionKeysMetas.size());
-            List<BKDReader.IteratorState> iteratorStates = new ArrayList<>(partitionKeysMetas.size());
-            try (NumericIndexWriter writer = new NumericIndexWriter(indexComponents, maxKeyLength, maxSSTableRowId, rowCount, false);
-                 IndexOutput bkdOutput = indexComponents.createOutput(IndexComponents.PRIMARY_KEYS);
-                 MetadataWriter metadataWriter = new MetadataWriter(indexComponents.createOutput(IndexComponents.GROUP_META));
-                 IndexOutput metaOutput = metadataWriter.builder(IndexComponents.PRIMARY_KEYS.name))
-            {
-                if (rowCount > 0)
-                {
-                    try (FileHandle segmentedPrimaryKeys = indexComponents.createFileHandle(indexComponents.PRIMARY_KEYS, true))
-                    {
-                        for (PartitionKeysMeta partitionKeysMeta : partitionKeysMetas)
-                        {
-                            iteratorStates.add(createIteratorState(partitionKeysMeta, segmentedPrimaryKeys.sharedCopy(), readers));
-                        }
-                    }
-                }
-                writer.writeIndex(bkdOutput, new MergeOneDimPointValues(iteratorStates, maxKeyLength)).write(metaOutput);
-            }
-            finally
-            {
-                indexComponents.deleteTemporaryComponent(indexComponents.PRIMARY_KEYS);
-                FileUtils.closeQuietly(iteratorStates);
-                FileUtils.closeQuietly(readers);
-            }
+            FileUtils.close(primaryKeys, primaryKeyOffsets, metadataWriter);
```

Comment:
Why use `close` here and not `closeQuietly`? Do we need to handle this?

Reply:
Well, there are some important things written on `close`, e.g. metadata. Generally I'd revert the question - why `closeQuietly`? Do we really want to not know about the failure to write? A failure to close can cause data loss.

Reply:
I agree but we always validate the files after the completion of writing anyway so if they are corrupt they will be picked up anyway. I believe that was the original reason for not handling it here. I am concerned about where this is now going to be handled and how that will interact with the validation and view management.

Reply:
In this case we don't want to throw any exceptions at all from these methods. But on the other hand - what about testing? This class is also used in testing. We definitely use this class to produce data files for benchmarking. 
Originally, before my changes, `complete` was marked as throwing `IOException`. So the callers should have been prepared to handle that.

Reply:
I'm not claiming consistency here and I'd like these things sorted out properly. I think that we should throw here but we just need to make sure that they are handled correctly during the write phase because we don't want to stop the memtable flush or compaction that is causing the write. If we have an error it needs to be handled in the least invasive way. As in, log the error, carry on and then fail validation so the user needs to rebuild the index.

Reply:
Looks like it is handled: 
https://github.com/datastax/cassandra/blob/dca62ddc6110ba04af73d8ce52da13a94ce48a5a/src/java/org/apache/cassandra/index/sai/disk/StorageAttachedIndexWriter.java#L191-L195

Reply:
I also checked that `nextRow` exceptions are also handled similarly in the SAIWriter.